orttraining/orttraining/models/runner/pipeline.cc

```diff
@@ -18,335 +20,618 @@ namespace onnxruntime {
+}
+void PipelineSlot::AddCompute(const int batch_id, const PipelineTask::Pass pass, const int upstream_time, const int upstream_stage) {
+  tasks_.push_back(PipelineTask{batch_id, PipelineTask::Type::Compute, pass, upstream_time, upstream_stage});
+}
+
+PipelineTask& PipelineSlot::operator[](int index) {
+  return tasks_[index];
+}
+
+const PipelineTask& PipelineSlot::operator[](int index) const {
+  return tasks_[index];
 }
 
-bool Slot::IsEmpty() const {
-  return type == Empty;
+PipelineTask& PipelineSlot::GetFrontAction() {
+  return tasks_.front();
 }
 
-bool Slot::IsForward() const {
-  return type == Forward;
+const PipelineTask& PipelineSlot::GetFrontAction() const {
+  return tasks_.front();
 }
 
-bool Slot::IsBackward() const {
-  return type == Backward;
+PipelineScheduler::PipelineScheduler(const int num_batches, const int num_stages) {
+  num_stages_ = num_stages;
+  num_batches_ = num_batches;
+  CreateComputeSchedule();
+
+  const size_t num_events_per_slot_compute_side = 2;
+  std::vector<int> compute_default_events(num_events_per_slot_compute_side, -1);
+  InsertEvents(compute_table_, num_events_per_slot_compute_side, compute_default_events);
+
+  CreateFullSchedule();
+
+  const size_t num_events_per_slot_side = 1;
+  std::vector<int> default_events(num_events_per_slot_side, -1);
+  InsertEvents(compute_commute_table_, num_events_per_slot_side, default_events);
 }
 
-void Slot::Show() const {
-  const int waited_event_id =
-      waited_events.size() > 0 ? waited_events[0] : -1;
-  const int waited_event_id_after_recv =
-      waited_events.size() > 0 ? waited_events[1] : -1;
-  const int recorded_event_id_before_send =
-      recorded_events.size() > 0 ? recorded_events[0] : -1;
-  const int recorded_event_id =
-      recorded_events.size() > 0 ? recorded_events[1] : -1;
-  switch (type) {
-    case Empty:
-      std::cout << "             ";
+// Print this structure following a fixed-length format.
+// It assumes there are at most 2 actions per slot.
+std::ostream& operator<<(std::ostream& stream, const PipelineSlot& slot) {
+  switch (slot.NumActions()) {
+    case 0:
+      // One empty action.
+      stream << "    ";
+      // Another empty action.
+      stream << "    ";
       break;
-    case Forward:
-      std::cout << " F" << batch_id << "("
-                << waited_event_id << "," << waited_event_id_after_recv << ","
-                << recorded_event_id_before_send << "," << recorded_event_id
-                << ")"
-                << " ";
+    case 1:
+      // Print an action.
+      stream << slot[0];
+      // One empty action.
+      stream << "    ";
       break;
-    case Backward:
-      std::cout << " B" << batch_id << "("
-                << waited_event_id << "," << waited_event_id_after_recv << ","
-                << recorded_event_id_before_send << "," << recorded_event_id
-                << ")"
-                << " ";
+    case 2:
```

Comment:
>case 2: [](start = 4, length = 7)

To make it a bit more robust, maybe add a default and assert it will never have more than 2 actions?

Reply:
Thank you. I will do it when refactorizing this class.

---
In reply to: [456118801](https://github.com/microsoft/onnxruntime/pull/4524#discussion_r456118801) [](ancestors = 456118801)